lib/rules/base.js

```diff
@@ -48,57 +43,300 @@ module.exports = function(options) {
+    // appropriate functions. Handlers get added to the before/after list
+    // according to whether they will be called before/after the rule's AST
+    // event handlers.
+    var astHandlers = {
+      CommentStatement: {
+        enter: { before: [], after: [] },
+        exit: { before: [], after: [] }
+      },
+      MustacheCommentStatement: {
+        enter: { before: [], after: [] },
+        exit: { before: [], after: [] }
+      },
+      ElementNode: {
+        enter: { before: [], after: [] },
+        exit: { before: [], after: [] },
+        keys: {
+          children: {
+            enter: { before: [], after: [] },
+            exit: { before: [], after: [] }
+          },
+          comments: {
+            enter: { before: [], after: [] },
+            exit: { before: [], after: [] }
+          }
+        }
+      }
+    };
+    // Keep a stack of ancestor elements so that when we encounter a comment
+    // that is the child of an element (i.e. not within the element's tag)
+    // we know what it's parent is so we can set up a listener for when we
+    // leave the comment's siblings.
+    var elementStack = [];
+    // We don't traverse in-element comments until after we've entered the
+    // element, but we need to apply such instructions before we enter the
+    // element. So, we have an element enter handler that processes that
+    // element's comments, which means we need to distinguish in-element
+    // comments from child-of-element comments during the traverse so we can
+    // ignore the former, but still process the latter.
+    var inElementComments = false;
+
+    //
+    // Wrap all visitors supplied by the rule, plus make sure we have entries
+    // for the visitors we use internally for instruction processing. This is
+    // done for two reasons -- one is to support instruction processing, but
+    // the other is to allow the rule's visitor handlers to have their this
+    // context set to the plugin, which isn't what HTMLBars does.
+    //
     for (var key in ruleVisitors) {
-      ruleVisitor = ruleVisitors[key];
-
-      visitors[key] = this.processVisitor(ruleVisitor);
+      visitors[key] = this._wrapVisitor(ruleVisitors[key], astHandlers[key]);
     }
 
-    var commentVisitor = visitors['CommentStatement'];
-    if (!commentVisitor) {
-      commentVisitor = visitors['CommentStatement'] = { enter: function() {} };
-    }
+    visitors.CommentStatement = visitors.CommentStatement || this._wrapVisitor(null, astHandlers.CommentStatement);
+    visitors.MustacheCommentStatement = visitors.MustacheCommentStatement || this._wrapVisitor(null, astHandlers.MustacheCommentStatement);
+    visitors.ElementNode = visitors.ElementNode || this._wrapVisitor(null, astHandlers.ElementNode);
+
+    //
+    // Set up our handlers
+    //
+
+    // Manage elementStack
+    astHandlers.ElementNode.keys.children.enter.before.push(function(elementNode) {
```

Comment:
Hmm...yeah, I suppose the outer handlers are static (although they dynamically push additional handlers). Conceptually I was thinking of it more as an observable mixin, where declaring the `astHandlers` is the equivalent of the mixin setting up its observable framework, and then these registrations are the equivalent of the subclass registering its handlers in a constructor.

I was actually going to set up `astHandlers` in the plugin constructor before realizing that they if the plugin instance were used for multiple traverses (which I don't think they are?) that the astHandlers would need to be reset for each traverse.

So anyway, I'm not that's actually an answer, but I think to my eye the code ends up more readable in this imperative form, rather than being declarative. But I could easily be swayed...